### PR TITLE
Add Transformers v5 compatibility

### PR DESCRIPTION
**Pull request by remidesbois1**
## What does this PR do?

Adds compatibility with Transformers v5 while preserving support for the existing Transformers v4 range.

## Changes

- Adds a compatibility layer for Transformers v5 API and model/config differences.
- Updates ONNX export, model patching, ONNX Runtime modeling, and pipelines.
- Keeps the supported Transformers range at `>=4.36,<6.0`.
- Updates documentation and CI coverage for the v5 compatibility target.

## Validation

- `ruff format --check .`
- `ruff check .`
- `git diff --check`
- Additional Transformers v4/v5 regression tests were run locally.

I'm open for suggestions.

### CODE REVIEW

**Review by DingmaomaoBJTU — APPROVED** [review]

**Comment by remidesbois1** [comment]
I temporarily added a few CI-only changes on my fork to reduce parallelism and Hub request pressure while validating the full runtime/export test matrix with the GH runners. Those changes were only intended to make the fork validation practical and have now been reverted from the PR branch (the force-pushes).

All other test jobs completed successfully. The only remaining failing exporter CLI job did not fail because of an export or inference regression: the run exhausted the Hugging Face Hub API quota and then received repeated 429 Too Many Requests responses while calling list_repo_files for the test models (0/500 requests remaining in the 300-second window). Once the rate limit was reached, the following parameterized cases failed immediately for the same reason.

The affected test suite passes locally on my machine, where the Hub rate limit is not hit. So far I have not found a functional regression in the Transformers v5 compatibility changes; the outstanding CI failure is caused by Hub request throttling on the GitHub-hosted runner.

I still found some things that needs to be "less monkey-patchy" but I'm fairly happy with the current result. I might also refactor the git history to split a 1200 lines addition into more than 2 commits.

**Comment by yxtay** [comment]
Is this still a draft PR? Ready to merge? Would appreciate transformer v5 support.

Closes #114.

**Comment by remidesbois1** [comment]
@yxtay 
> Is this still a draft PR? Ready to merge? Would appreciate transformer v5 support.

I'd also love to see Transformers v5 support merged. 🙂

I'm not a maintainer though, so I don't have any control over when (or if) this gets merged.

I still have a few things I'd like to clean up to remove some "per model patch" and easier to maintain, but the overall compatibility layer is working well. Once I'm happy with those final refinements, I'll mark the PR as ready for review and let the maintainers decide the next steps.

If you want you can already pull my dev branch and try it yourself, I'd love a feedback.

**Comment by remidesbois1** [comment]
Also, for compatibility with some models, I had to re-implement old/depreciated transformers v4 method in optimum/transformers_compat.py : 

- head masking
- head pruning
- model parallelism

Which I mostly reinjected into PreTrainedModel, pytorch_utils and sys.modules ; 
This might not be the correct solution to support those old models that are not supported anymore by transformers v5.

Two options that I found are either : 
- Raise RuntimeError saying that the model relied on Transformers v4 APIs that were removed.
- have a context manager like this : 
```
@contextmanager
def legacy_transformers_v4_remote_code_compatibility():
    install_legacy_transformers_shims() # load the helpers reimplementation

    try:
        yield
    finally:
        remove_legacy_transformers_shims() # unload them
```

with a cli option `--allow-legacy-transformers-code`

But this would be strange when using this with transformers v4 installed, this need to be discussed, maybe supporting transformers v4 as well as v5 isn't the best idea.

There are other things that might need to be discussed like the restauration of the pipelines : 
- text2text-generation
- summarization
- translation
- image-to-text

---

For existing trust_remote_code repositories, it also restores missing v4 configuration defaults such as:
```
torchscript = False
pruned_heads = {}
tie_word_embeddings = True
is_decoder = False
add_cross_attention = False
bos_token_id = None
eos_token_id = None
```

---

Recent PyTorch versions may reject a dimension declared as fully dynamic when tracing proves it to be static.

I changed dynamic axes that occur only once from Dim.DYNAMIC to Dim.AUTO.
This allows some exports to succeed, but it can also mean that a dimension declared dynamic in an OnnxConfig becomes static in the resulting export.
Do we : 
- prefer a successful export with less dynamism than requested
- warn when a dimension becomes static
- fail when the declared dynamic-shape contract cannot be preserved.

---

So this is absolutely not ready for merge even through it "work", there are still some choices to be made. And I'm likely not the one making them 😊

**Comment by dsubunow** [comment]
Hi @remidesbois1 any context/thoughts on when this will be reviewed by a maintainer? 
I also do see: https://github.com/huggingface/optimum-onnx/pull/121
